### PR TITLE
Ultra Merge

### PR DESCRIPTION
**Pull request by Gamenot**

### CODE REVIEW

File: ultra/baselines/ddpg/ddpg/policy.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
+# code iintegrated from
+#  1- https://github.com/udacity/deep-reinforcement-learning
+#  2- https://github.com/sfujim/TD3/blob/master/TD3.py
```
Comment (Gamenot) [inline ultra/baselines/ddpg/ddpg/policy.py:3]: Have the licenses been covered for these files, I see they are both MIT which means we need to carry on their copyright notice.

File: ultra/baselines/ddpg/ddpg/noise.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
+"""
+source code: https://github.com/ShangtongZhang/DeepRL/deep_rl/component/random_process.py
```
Comment (Gamenot) [inline ultra/baselines/ddpg/ddpg/noise.py:2]: Are the MIT license requirements fulfilled?

File: ultra/baselines/ppo/ppo/policy.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
+# some parts of this implementation is inspired by https://github.com/Khrylx/PyTorch-RL
```
Comment (Gamenot) [inline ultra/baselines/ppo/ppo/policy.py:1]: Are the MIT license requirements fulfilled?

File: ultra/docker/remote.sh [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+# An example to execute this script:
+# source remote.sh <user name> <password> <server name> <container name>
+# source remote.sh z84216771 abcd1234 GX3 ultratest
+
+# Servers
+# CX3 = 10.193.241.233
+# CX4 = 10.193.241.234
+# Compute-4 = 10.193.192.17
+# Compute-11 = 10.193.192.113
+# GX3 = 10.193.241.239
```
Comment (Gamenot) [inline ultra/docker/remote.sh:12]: Remove

File: ultra/docker/remote.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        ;;
+    "GX3")
+        ADD=10.193.241.239
+        DST=/data/research
+        GPUFLAG="--gpus=all"
+        ;;
+    *)
+        echo "Unknown server"
+        return 5
+        ;;
```
Comment (Gamenot) [inline ultra/docker/remote.sh:49]: This should not be in here I think

File: ultra/env/__init__.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+# register(
+#     id="task1",
+#     entry_point="ultra.env.ultra_env:UltraEnv",
+# )
+#
+# register(
+#     id="task2",
+#     entry_point="ultra.env.ultra_env:Task2Env",
+# )
```
Comment (Gamenot) [inline ultra/env/__init__.py:20]: Are these going to be used? Otherwise these comments should be cleaned up.

File: ultra/setup.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    zip_safe=True,
+    python_requires=">=3.7",
+    install_requires=[
+        "smarts@git+https://github.com/huawei-noah/SMARTS.git",
+        "setuptools>=41.0.0,!=50.0",
+        "dill",
+        "black==20.8b1",
+    ],
+    dependency_links=[
+        # "smarts@git+https://gitlab.smartsai.xyz/smarts/SMARTS/-/releases/v0.4.1"
```
Comment (Gamenot) [inline ultra/setup.py:23]: Remove

File: ultra/tests/evaluate_test.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
+# import unittest, ray, glob, gym, os
```
Comment (Gamenot) [inline ultra/tests/evaluate_test.py:1]: Is this file intended to be commented out?

File: ultra/utils/common.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+def rotate2d_vector(vectors, angle):
+    ae_cos = np.cos(angle)
+    ae_sin = np.sin(angle)
+    rot_matrix = np.array([[ae_cos, -ae_sin], [ae_sin, ae_cos]])
+    vectors_rotated = np.inner(vectors, rot_matrix)
+    return vectors_rotated
+
+
+def plot_experiment_data(experiment_path, save_dir):
+    # logs/experiment-2020.9.25-12:16:14/pkls/Train
```
Comment (Gamenot) [inline ultra/utils/common.py:22]: Could this be misleading?

File: cli/ultra.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+import glob
+import shutil
+import subprocess
+from pathlib import Path
+
+import click
+from rich import print
+
+
+@click.group(name="ultra")
```
Comment (liamchzh) [inline cli/ultra.py:31]: I think we don't need cli/ultra.py for now because  what the cli/zoo.py doing is building the agents so that we can use them as 3rd party libs in scenarios.

File: ultra/baselines/adapter.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
+import numpy as np
+from scipy.spatial import distance
+import random, math
+from sys import path
+
+path.append("./ultra")
```
Comment (liamchzh) [inline ultra/baselines/adapter.py:6]: Maybe put this line after import?

File: ultra/docker/remote.sh (hunk 5 above)
Comment (adreena) [reply ultra/docker/remote.sh:49]: @Adaickalavan

File: ultra/tests/evaluate_test.py (hunk 8 above)
Comment (adreena) [reply ultra/tests/evaluate_test.py:1]: it was throwing errors in the previous version so I commented out but we're gonna update these later.
so im gonna remove test folder for now and keep it just locally

File: ultra/baselines/adapter.py (hunk 11 above)
Comment (adreena) [reply ultra/baselines/adapter.py:6]: why? the code works

File: ultra/utils/common.py (hunk 9 above)
Comment (adreena) [reply ultra/utils/common.py:22]: we used this module for plots in the paper, why is this misleading?

File: ultra/setup.py (hunk 7 above)
Comment (adreena) [reply ultra/setup.py]: Im gonna remove setup.py

File: ultra/env/__init__.py (hunk 6 above)
Comment (adreena) [reply ultra/env/__init__.py:20]: no, good catch. it was meant for a proposal and I forgot to remove them

File: ultra/README.md [hunk 12]
```diff
@@ -0,0 +1,124 @@
+[![coverage report](https://gitlab.smartsai.xyz/smarts/ULTRA/badges/master/coverage.svg)](https://gitlab.smartsai.xyz/smarts/ULTRA/-/commits/master)
+[![pipeline status](https://gitlab.smartsai.xyz/smarts/ULTRA/badges/master/pipeline.svg)](https://gitlab.smartsai.xyz/smarts/ULTRA/-/commits/master)
```
[the remark this answers is not in the document]
Comment (adreena) [reply ultra/README.md:2]: is it possible to show coverage badge per folders? like for ultra we keep our own badge?

**Comment by adreena** [comment]
@christianjans @JenishPatel99 did you guys plan to update some parts of the README.md for this sprint?

File: ultra/docker/remote.sh (hunk 5 above)
Comment (adreena) [reply ultra/docker/remote.sh]: I removed the docker folder from git

File: ultra/README.md [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+- To map additional Docker container volumes such as `ULTRA/ultra/scenarios/task1` to local storage, edit `ULTRA/ultra/docker/remote.sh` to include `--volume=${DST}/ultra/scenarios/task1/:/ULTRA/ultra/scenarios/task1/`.
+- Some regular commands
+  ```sh
+  # Copy folder into remote server
+  $ sshpass -p <password> scp -r <folder to copy> <username>@<server address>:<destination path>
+  $ sshpass -p abcd1234 scp -r ${PWD}/ultra/scenarios/task3 z84216771@10.193.241.239:/data/research/ultra/scenarios/
+
+  # Login into remote server
+  $ sshpass -p <password> ssh <username>@<server address>
+  $ sshpass -p abcd1234 ssh z84216771@10.193.241.239
```
[the remark this answers is not in the document]
Comment (adreena) [reply ultra/README.md]: these were meant to run for internal gpu server, im removing them

File: ultra/docker/remote.sh (hunk 5 above)
Comment (Adaickalavan) [reply ultra/docker/remote.sh]: The documentation for code deployment in our internal servers, has been uploaded into our own GitLab. I have forwarded the link in our chat group. Hence, please feel free to remove the above file.

File: ultra/utils/common.py (hunk 9 above)
Comment (Gamenot) [reply ultra/utils/common.py:22]: Sorry it was just a nit, I should have been specific, I meant the comment.

**Comment by christianjans** [comment]
@adreena sorry actually I think we will update parts of the README and other documentation once this branch is merged.

File: ultra/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Gamenot) [reply ultra/README.md]: A badge should be possible but it may need some customisation to get it to work as intended.

File: ultra/baselines/ppo/ppo/policy.py (hunk 3 above)
Comment (adreena) [reply ultra/baselines/ppo/ppo/policy.py:1]: @Gamenot Could you ask Nhat? he was dealing with licenses back in November

File: ultra/docker/remote.sh (hunk 4 above)
Comment (adreena) [reply ultra/docker/remote.sh]: entire docker/ folder is removed

File: ultra/baselines/ddpg/ddpg/noise.py (hunk 2 above)
Comment (adreena) [reply ultra/baselines/ddpg/ddpg/noise.py:2]: please ask Nhat, he was fixing the licenses

File: ultra/baselines/ddpg/ddpg/policy.py (hunk 1 above)
Comment (adreena) [reply ultra/baselines/ddpg/ddpg/policy.py:3]: Ask Nhat.

File: cli/ultra.py (hunk 10 above)
Comment (adreena) [reply cli/ultra.py:31]: @liamchzh what if zoo is not a part of smarts in future?

File: ultra/train.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        "--seed", help="environment seed", default=2, type=int,
+    )
+    args = parser.parse_args()
+
+    num_cpus = max(
+        1, psutil.cpu_count(logical=False) - 1
+    )  # remove `logical=False` to use all cpus
+
+    policy_class = "ultra.baselines.sac:sac-v0"
+    # ray_kwargs = default_ray_kwargs(num_cpus=num_cpus, num_gpus=num_gpus)
```
Comment (AlexLewandowski) [inline ultra/train.py:134]: Should this line be commented out, or can users set resources and other ray kwargs elsewhere?

File: ultra/baselines/common/social_vehicle_extraction.py [hunk 15]
```diff
@@ -0,0 +1,270 @@
...
+        social_vehicles_permuted = social_vehicles_permuted[:social_capacity]
+
+    return social_vehicles_permuted
+
+
+def make_fake_social_vehicle(
+    ego_vehicle_pos, ego_vehicle_heading, prev_ego_wp, max_dist_social_vehicle
+):
+    # DG: I changed to zeros because I think it'll be easier for learning
+    return [0.0, 0.0, 0.0, 0.0]
```
Comment (AlexLewandowski) [inline ultra/baselines/common/social_vehicle_extraction.py:212]: Is this the new default or will it be reverted to the commented version below?

**Review by AlexLewandowski — APPROVED** [review]

**Review by christianjans — APPROVED** [review]

**Comment by JingfeiPeng** [comment]
To add headers or check if headers exist on ultra files
Generate headers run:  `make gen-header`
Test if headers exist: `make header-test`

Though the script needed to be updated to include the ultra folder

File: ultra/train.py (hunk 14 above)
Comment (adreena) [reply ultra/train.py:134]: @AlexLewandowski it wasn't compatible with ray 0.8, so I commented it out

File: ultra/baselines/common/social_vehicle_extraction.py (hunk 15 above)
Comment (adreena) [reply ultra/baselines/common/social_vehicle_extraction.py:212]: it's supposed to fill the empty spots for social vehicles, say if we have 20 vehicles as the input of the network but there are not enough vehicles in the range we fill them with 0

File: ultra/setup.py (hunk 7 above)
Comment (AlexLewandowski) [reply ultra/setup.py]: ultra/Dockerfile needs to be updated, it still references setup.py on line 115

**Review by adreena — APPROVED** [review]
@Gamenot please go ahead an merge ultra after tomorrow's standup if there are no more questions

**Comment by Gamenot** [comment]
This will be merged into master today.

File: ultra/baselines/common/replay_buffer.py [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+# Do not make any change to this file when merging. Just use my version.
```
Comment (Gamenot) [inline ultra/baselines/common/replay_buffer.py:22]: This is fairly opinionated.

Comment (adreena) [reply ultra/baselines/common/replay_buffer.py:22]: was this added through the gen_header?

Comment (Adaickalavan) [reply ultra/baselines/common/replay_buffer.py:22]: This line appears to be committed by @adreena on 18 Sept 2020, with commit message `first commit` and commit hash 67fd489, in the GitLab repository.
